share/gpodder/extensions/youtube-dl.py

```diff
@@ -315,7 +315,9 @@ def add_format(self, gpodder_config, opts, fallback=None):
     def fetch_video(self, url, tempname, reporthook):
         opts = {
             'paths': {'home': os.path.dirname(tempname)},
-            'outtmpl': os.path.basename(tempname) + '.%(ext)s',  # Postprocessing breaks without ext
+            # Postprocessing in yt-dlp breaks without ext
+            'outtmpl': os.path.basename(tempname) + (
+                '.%(ext)s' if program_name == 'yt-dlp' else ''),
```

Comment:
`outtmpl` should be `(os.path.basename(tempname) if program_name == 'yt-dlp' else tempname) + '.%(ext)s'`. The `paths` is fine to always send because youtube-dl ignores it, and that means youtube-dl needs the full tempname path, not basename.

I don't know if youtube-dl post-processing breaks but it does support %(ext)s. I can't test it right now, do you see any reason to not append the extension for youtube-dl?

Reply:
I tested this with recent versions of both yt-dlp and youtube-dl. Both download fine with `.%(ext)s` added, but with youtube-dl, the preview functionality in gPodder (playing the temp-file being downloaded) is broken when the ext is added.

The preview works in youtube-dl only with single-file formats. With yt-dlp, the ext needs to be added, otherwise merging multi-file formats and other post-processing will fail. Is there an easy way to disable the preview for files downloaded with custom dowloader (i.e. the youtube-dl extension)?

Reply:
Preview is broken with youtube-dl and yt-dlp for me, because it passes "{name}.{ext}.partial" to the video player for both. https://github.com/gpodder/gpodder/blob/master/src/gpodder/model.py#L627 produces this filename and maybe custom downloaders could get a new method to resolve partial filenames. It could append the extension if known or return None if the partial file shouldn't be previewed. The preview menu could call get_playback_url and disable Preview if None.

If retrieve_resume could get the extension before downloading, it could store it in the downloader and we could have DownloadTask.run() make its `downloader` variable available in the task so `episode.download_task.custom_downloader.extension` is accessible by this new partial filename resolver.

If possible, retrieve_resume should detect DASH formats and set a flag so resolving the partial filename always returns None. Or we simply disable preview for all youtube-dl/yt-dlp episodes by always returning None.

Reply:
The preview for the youtube-dl extension downloads is now implemented more or less as @auouymous describes above. Preview should now work for normal downloads and yt-dlp and youtube-dl downloads which do not use a fragment downloader (DASH etc.). Needs at least some testing and squashing before merging.